src/pnpq/apt/connection.py

```diff
@@ -0,0 +1,358 @@
+    # connection is held by a single object that cannot be
+    # closed. However, we should still probably implement a context
+    # manager for this object.
+    #
+    # The tricky part is managing the threads; they will all need to
+    # be paused before closing the serial connection... and then how
+    # do we clean up the child threads if this object gets cleaned up?
+
+    # def __enter__(self) -> None:
+    #     self.open()
+
+    # def __exit__(self, exc_type, exc_value, exc_tb) -> None:
+    #     self.close()
+
+    # def open(self) -> None:
+    #     self.connection.open()
+
+    # def close(self) -> None:
+    #     self.connection.flush()
+    #     self.connection.close()
+
+    def rx_dispatch(self) -> None:
+        with self.rx_dispatcher_thread_lock:
+            while True:
+                partial_message: None | AptMessageForStreamParsing = None
+                full_message: Optional[AptMessage] = None
+                try:
+                    message_bytes = self.connection.read(6)
+                    partial_message = AptMessageForStreamParsing.from_bytes(
+                        message_bytes
+                    )
+                    message_id = partial_message.message_id
+                    if partial_message.data_length != 0:
+                        message_bytes = message_bytes + self.connection.read(
+                            partial_message.data_length
+                        )
+                    if partial_message.message_id in AptMessageId:
+                        message_id = AptMessageId(partial_message.message_id)
+                        full_message = getattr(
+                            pnpq.apt.protocol, f"AptMessage_{message_id.name}"
+                        ).from_bytes(message_bytes)
+                        assert isinstance(full_message, AptMessage)
+                        self.log.debug(
+                            event=Event.RX_MESSAGE_KNOWN,
+                            message=full_message,
+                        )
+                        with self.rx_dispatcher_subscribers_lock:
+                            for queue in self.rx_dispatcher_subscribers.values():
+                                queue.put(full_message)
+                    else:
+                        # Log and discard unknown messages
+                        self.log.debug(
+                            event=Event.RX_MESSAGE_UNKNOWN,
+                            message=partial_message,
+                            bytes=message_bytes,
+                        )
+                # TODO this is too general, do not catch Exception
+                except Exception as e:  # pylint: disable=W0718
+                    self.log.error(
+                        event=Event.UNCAUGHT_EXCEPTION,
+                        exc_info=e,
+                        partial_message=partial_message,
+                        full_message=full_message,
+                    )
+
+    def tx_poll(self) -> None:
```

Comment:
So, by default, `tx_poll` probably shouldn't do anything. At least, it shouldn't send any messages.

So, we have a few options for how to solve this.

1. **Remove the `tx_poll` function, code, and thread completely from connection.py**: This is the easiest to do. We would move this part of the code into the device implementation classes.
2. **Provide a default no-op implementation in connection.py**: In this case, we would need to create a subclass of `AptConnection`, called, for instance, `StatusUpdatingAptConnection`, to implement polling as we think is appropriate.
3. **Pass in a `Callable` to the `AptConnection` constructor**: In other words, require a function to be passed in that `AptConnection` would run inside the `while True` loop, replacing lines 241 to 254.

Of the three options, I think that option 1 is probably the best. It gives us flexibility in what we can poll for and how we handle the results. One downside is that it means we have to do thread management in more places.

Option 2 gets tricky if we want to share information from the polling with the driver class.

Option 3 _could_ let us share information with the driver class, but in order to do so we would have to change the way that we initialize the connection and the driver object. And we would still have to deal with thread synchronization in the driver class.

Reply:
I see that different devices could implement polling slightly differently. But I thought the main point of this class is so that all the threading can happen separately. If we bring `tx_poll` back then we only abstracted 2/3 of the threading. It's definitely the quicker way to do this if we don't want to spend longer on this particular task tho.

Reply:
@TomShen1234 Another way of thinking about this is that the `AptConnection` class abstracts out the concept of an APT connection, e.g., the process of sending and receiving APT messages. It doesn't have anything to do with what messages should be sent using that connection.

For that matter, the fact that it uses threads is an implementation detail and not part of the public API.

Another consideration is that we might not always need polling. In theory, we can use `HW_START_UPDATEMSGS` to instead get the device to send messages to us automatically. The documentation on the `ACK` messages is also very confusing, and it's not clear if we really need them or not.

My inclination right now is still to make the polling part of the device driver class, not the connection class.